lib/cloudinstancecredentials/azurecredentials.py

```diff
@@ -19,7 +19,7 @@
 from cloudinstancecredentials.instancemetadata import InstanceMetadata
 
 
-def detect_framework():
+def has_metadata_access():
```

Comment:
This should now be in the factory executed unconditionally inside the try i.e.
```
try:
    azure access
    return azure-instance
except Exception:
    pass
try:
    ec2 access
    return ec2-instance
....
```

Reply:
I'd prefer to use the boolean validation rather than rely on an exception being raised. 

Both Azure and AWS use the same IP address for metadata service, so _requests_ will succeed in getting http://169.254.169.254 on either framework. To be sure I'm in Azure, I have to get a specific endpoint that only Azure would respond to, _and_ have the response code be a 200. In AWS, the same request would succeed (e.g. no exception raised by _requests_) but raise a 404 response code instead.

In order to use the `try` validation as you suggested, I would have to artificially raise an exception in both Azure & AWS validators. I'd rather rely on the natural boolean response, and handle an unexpected exception in the check, rather than artificially add an exception and then handle it. I also prefer the readability of the `if` version... it just makes more sense in my head.

Reply:
The try is already in the code, I am not suggesting to change that. The try will have to stay in case something else goes wrong and requests throws an exception, just take the implementation from here and stick it into the factory and let it execute unconditionally.

Reply:
Okay, I get it now. Thinned down in d914468.